### DISCUSSION TITLE
where is the file uni_dmft.py

### DISCUSSION THREAD

**Discussion by Lanlin123** [TRIQS/solid_dmft#45] [Q&A] [answered]
Dear TRIQS/solid_dmft developers,

In the tutorial for PrNiO3, I can't find the file "uni_dmft.py" to start the CSC dmft caculation. I instead use "mpirun -n 32 solid_dmft“ to start the process. However it leads to the error as shown in the following figure. I wonder where I can get access to the file "uni_dmft.py". For the error, Can you give some advice? 

![image](https://user-images.githubusercontent.com/51743630/212327946-4602fd98-93a6-4267-82c2-324a8e3a9b17.png)

Best regards,
Lanlin.

**Comment by the-hampel** [comment]
Hi @Lanlin123,
are you talking about this tutorial: https://triqs.github.io/solid_dmft/tutorials/PrNiO3_csc_vasp_plo_cthyb/tutorial.html ? Indeed this is mentioned there. I will correct it. `uni_dmft.py` is the old name of the software - you can just use `solid_dmft` instead. 

Which version of solid_dmft are you using currently? I know the error but it has been fixed a few month back: https://github.com/TRIQS/solid_dmft/commit/e0479ab10d7af7d1ef96829d0e92880ff0c6947a . You can use the stable release branch https://github.com/TRIQS/solid_dmft/commits/3.1.x and everything should work. For pip we did not push yet a release to fix this. We will do so shortly. 

Thanks for reporting,
Alex

**Reply by the-hampel** [reply]
`uni_dmft.py` mentioning in PrNiO3 tutorial fixed by c9888e96c45eb6ae9c3181451e1ea9bfef204aec

**Reply by the-hampel** [reply]
I created a release 3.1.3 that should address your problem: https://github.com/TRIQS/solid_dmft/releases/tag/3.1.3 . This is also available now via pip.

**Reply by Lanlin123** [reply]
Hi Alex,

Sorry late to see your reply. Thank you so much for your help. I have pip upgrate solid_dmft to version 3.1.3, but there exists another error shown below.
![image](https://user-images.githubusercontent.com/51743630/219044064-811cba6c-5596-401d-beb6-31f8303e6386.png)

Best,
Lanlin.

**Reply by the-hampel** [reply]
Hi @Lanlin123,
ah yes you are right I made a mistake when creating the package. I forgot the version.py module in the latest release. The rest of solid_dmft should work without problems, but to run it I have to fix it. I will fix it ASAP.

**Reply by the-hampel** [reply]
I just updated the pypi package to ver 3.1.4 which fixes the problem https://pypi.org/project/solid-dmft/3.1.4/ . Just update solid_dmft via `pip install --upgrade solid_dmft` .  Let me now if there are further problems.

**Reply by Lanlin123** [reply]
Hi Alex,

Thank you so much for your help! Now it works well. But when I used plot_correlated_bands.py to plot A(k,w), another  error occured, as shown below.

![image](https://user-images.githubusercontent.com/51743630/221733285-8dacbfae-c654-4203-8a56-ac81d90d7a28.png)

Can you give me some advice for this problem? Thanks!

Best,
Lanlin.

**Reply by the-hampel** [reply]
Hi Lanlin,
good to hear that everything works now. 

Unfortunately the spectral function post-processing and plotting routines available in solid_dmft right now are limited to be used with the wannier90 interface without bloch_mode. This is because the calculation of the spectral function needs the non-interacting input dispersion eps(k) at arbitrary k-points. 

The error tells me that you tried to use the plotting routines with a projector calculation.

Right now the construction of the lattice Green's function (for the spectral function) is done in solid_dmft using Fourier transforms of the real space Wannier Hamiltonian. When using projectors, i.e. via Vasp LOCPROJ flag or QE wannier90 bloch mode, this is not possible. We will work on the possibility to use also the projector mode with the postprocessing routines in the future. For now you would either need to switch to the wannier90 interface, or manually calculate the proj_mat first on the high symmetry k-point lines and then manually construct the lattice Green's function.

- Alex

**Reply by Lanlin123** [reply]
Hi Alex,

Thank you for your reply. I wonder how to set the wannier90 interface without bloch_mode in csc calculation. I guess setting write_u_matrices = .false. in seedname.win. Is it true?

Best,
Lanlin.

**Reply by the-hampel** [reply, marked as answer]
This is a bit tricky. For this you have to carefully check your system. Are you performing the calculation for PrNiO3 as described in the tutorial? You have to make sure that your system does not have any disentanglement (more input DFT bands than Wannier states). If this is the case you can run after the DFT+DMFT calculation is finished once the Wannier converter manually with setting `bloch_basis=False` :
```
Converter = Wannier90Converter(seedname=yourseedname, rot_mat_type='hloc_diag', bloch_basis=False)
Converter.convert_dft_input()
```
Running the converter will replace some things in the dft_input group in the h5 archive, so make a copy of the result h5 archive first. Afterwards you can use the new h5 file to pass to `get_dmft_bands` from the postprocessing routines. You should carefully check that there was no disentanglement in the system as I mentioned. Otherwise the result will be not correct.

**Reply by Lanlin123** [reply]
Hi,
 
Recently, I try to use solid_dmft to calculate paramagnetical NiO as in triqs/dft_tools tutorial. However I can't get correct result, i.e., I can't open the Mott gap from the DFT metal state. I have used 5 Ni d orbitals and 3 O p orbitals near the Fermi level to construct the wannier Hamitonian and chosen these d orbitals as correlated ones. The Ni2+ has 8 d electrons and O2- 6 p electrons. I solved the impurity model by hubbardI and cthyb solvers, but their results both kept in a metal state. The files I used could be seen from https://github.com/Lanlin123/NiO/ . These ipynb files include my results. Could you please give me some advice? Thank you so much.

Best regards,
Lanlin.

**Reply by Lanlin123** [reply]
Hi,

I found hubbardI solver could open a gap just like:
![image](https://user-images.githubusercontent.com/51743630/232290066-0c40a907-5915-4443-a160-cc2b4ca6757b.png)
which however deviated from data from https://journals.aps.org/prb/abstract/10.1103/PhysRevB.74.195114. And the k-resolved spectral function plotted from PCB script is like :
![image](https://user-images.githubusercontent.com/51743630/232290273-0ec6b650-f209-442b-88e3-8d9bd67e1c0e.png)
which is weird. I gauss the chemical potential is not correct in PCB script. Then my result from cthyb is weird too, shown like:
![image](https://user-images.githubusercontent.com/51743630/232290764-18cee7e9-ed3e-4f1b-a686-aeb6fc104075.png)
The spectral weight tends to approaches zero at the ω=0 point, but it's not easy to improve the parameters to let it work.

**Comment by the-hampel** [comment]
Hi @Lanlin123,
sorry for the slow response. 

I had a look at your ipynb notebooks and I think the results look okay. Why do you say the chemical potential in the Akw is not correct? It says 
```
μ=14.6876 eV set for calculating A(k,ω)
```
and I think this is correct. You wannier90 _hr.dat file has a chemical potential of ~15.51 and your DMFT mu is roughly -1 eV. Since you subtracted mu in the w90_converter this makes sense to me. You correctly set `add_mu_tb=True` in the get_dmft_bands routine. Let's first look at the HI solver. This is faster to rerun. Could you tell me what the value of the DC potential is, that solid_dmft reports? There is somewhere a line reporting something like calculating DC and then the DC pot is reported. The gap opening in NiO is very sensitive to this value so maybe you are not close to the value that opens a gap? The value needs to be close to 59 eV.

Best,
Alex

**Reply by Lanlin123** [reply]
Hi Alex,

Thank you so much for your reply. I am so sorry to miss your message in the previous days. In the case of my uploaded ipynb notebooks, I use **dc_dmft = False** in my input file. So I have only one DC potential value, which is 61.947058 eV. By the way, as the previous papers suggest that O 2p orbitals are hybridized with Ni 3d orbitals, HubbardI method may not describe NiO well.  

Best,
Lanlin

**Reply by the-hampel** [reply]
Could you try to run a calculation adding the following to your dmft_config.ini file:
```
[advanced]
dc_fixed_value = 59.0
```
and see if that changes anything in your resulting spectral function. But you also tried with the cthyb impurity solver right? That should work at least.

Best,
Alex

**Reply by Lanlin123** [reply]
Thank you for your reply. Yeah, I have done a cthyb calculation setting _dc_fixed_value = 59.0_  as you suggested. It works as a gap occuring in DOS. Please see my input, result and log files from https://github.com/Lanlin123/NiO_fixDC-59 .  Although the gap is much smaller than the experimental one and the k-resolved spectral looks not so good near the Fermi Surface. And I wonder, if the result depends on the DC value so much, how could we choose a reasonable one when we use solid_dmft to calculate an undiscovered material. 

Best,
Lanlin

**Reply by the-hampel** [reply]
I think your results look reasonable. I think you could converge them better with more MC samples, but the result will not change qualitatively. 

The DC issue is one of the fundamental problems in any DFT++ methods (same for DFT+U and DFT+DMFT), see for example dx.doi.org/10.1016/j.elspec.2010.05.021 or http://dx.doi.org/10.1103/PhysRevLett.115.196403 . Although there has been some promising work on finding the correct DC between DFT and embedding methods it is a intrinsically ill posed problem. Other methods such as GW+DMFT might be a solution. 

I can say that most of the time the FLL DC with nominal occupations give reasonable results. Or you try to work with a minimal low-energy model, without explicit oxygen-p orbitals. This however shifts the problem only around. 

Best,
AlexI think your results look reasonable. I think you could converge them better with more MC samples, but the result will not change qualitatively. 

The DC issue is one of the fundamental problems in any DFT++ methods (same for DFT+U and DFT+DMFT), see for example dx.doi.org/10.1016/j.elspec.2010.05.021 or http://dx.doi.org/10.1103/PhysRevLett.115.196403 . Although there has been some promising work on finding the correct DC between DFT and embedding methods it is a intrinsically ill posed problem. Other methods such as GW+DMFT might be a solution. 

I can say that most of the time the FLL DC with nominal occupations give reasonable results. Or you try to work with a minimal low-energy model, without explicit oxygen-p orbitals. This however shifts the problem only around. 

Best,
Alex

**Reply by Lanlin123** [reply]
Yeah, I see. Thank you so much!

Best,
Lanlin.
